app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/PeerGroupHandshakeAsyncTask.java

```diff
@@ -0,0 +1,158 @@
+package org.kiwix.kiwixmobile.zim_manager.local_file_transfer;
+
+import android.net.Uri;
+import android.net.wifi.p2p.WifiP2pInfo;
+import android.os.AsyncTask;
+import android.util.Log;
+
+import java.lang.ref.WeakReference;
+import org.kiwix.kiwixmobile.BuildConfig;
+
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.ArrayList;
+
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.FileItem.FileStatus.TO_BE_SENT;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.PEER_HANDSHAKE_PORT;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.getFileName;
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once two peer devices are connected through wifi direct, this task is executed to perform a
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver. Finally, the onPostExecute() of the sender
+ * initiates the file transfer through {@link SenderDeviceAsyncTask} on the sender and using
+ * {@link ReceiverDeviceAsyncTask} on the receiver.
+ */
+class PeerGroupHandshakeAsyncTask extends AsyncTask<Void, Void, InetAddress> {
+
+  private static final String TAG = "PeerGrpHndshakeAsyncTsk";
+  private final String HANDSHAKE_MESSAGE = "Request Kiwix File Sharing";
+
+  private WeakReference<LocalFileTransferActivity> weakReferenceToActivity;
+  private WifiP2pInfo groupInfo;
+
+  public PeerGroupHandshakeAsyncTask(LocalFileTransferActivity localFileTransferActivity, WifiP2pInfo groupInfo) {
+    this.weakReferenceToActivity = new WeakReference<>(localFileTransferActivity);
```

Comment:
[WeakReference should be the last resort to fix a memory leak.](https://medium.com/google-developer-experts/weakreference-in-android-dd1e66b9be9d)

Reply:
I had a design query: The article is essentially using a callback from what I understood, but instead of implementing the interface directly in the class (like how I have been using callbacks until now), it's returning (& setting) an anonymous listener object with the required implementation.

Are there situations where there is a preference given to either of them, or does it depend on the org's code guidelines?

Reply:
[Commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/e6fa97abba36b6d8d81b34d7484820fc066b015e)
Async tasks only deal with WDM now

Reply:
It would also be ideal if the Asyncs talked to WDM viw callback

Reply:
Anonymous callbacks I would use when the callbacks are very limited, take the example of DialogShower, in its usages I have performed the positive/negative actions at the call site.
When there are a large number and the express purpose of the class is to be a listener for the actions then it makes sense for it to implement them as it describes the functionality of the class

Reply:
Thanks for clarifying about the choice of implementation
> It would also be ideal if the Asyncs talked to WDM viw callback

Yes, I'll be doing that as well. LFTA & BroadcastReceiver are done (using callbacks), async-tasks are next :+1: